packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -679,11 +675,27 @@ export abstract class FluidDataStoreContext
 		this.channel!.setConnectionState(connected, clientId);
 	}
 
-	public notifyReadOnlyState(readonly: boolean): void {
+	public notifyReadOnlyState(): void {
 		this.verifyNotClosed("notifyReadOnlyState", false /* checkTombstone */);
 
-		this.channel?.notifyReadOnlyState?.(readonly);
-		this._contextDeltaManagerProxy.setReadonly(readonly);
+		// These two calls achieve the same purpose, and are both needed for a time for back compat
+		this.channel?.notifyReadOnlyState?.(this.isReadOnly());
+		this._contextDeltaManagerProxy.emitReadonly();
+	}
+
+	/**
+	 * Updates the readonly state of the data store based on the staging mode.
+	 *
+	 * @param staging - A boolean indicating whether the container is in staging mode.
+	 * If true, the data store is set to readonly unless explicitly allowed by its policies.
+	 */
+	public notifyStagingMode(staging: boolean): void {
+		// If the `readonlyInStagingMode` policy is not explicitly set to `false`,
+		// the data store is treated as readonly in staging mode.
+		if (this.channel?.policies?.readonlyInStagingMode !== false) {
+			this.isStagingReadOnly = staging;
+			this.notifyReadOnlyState();
+		}
```

Comment:
If you take the above suggestion, then this becomes
```suggestion
		const oldReadOnlyState = this.isReadOnly();
		this.isStagingMode = staging;
		if (this.isReadOnly() !== oldReadOnlyState) {
			this.notifyReadOnlyState();
		}
```

Whether you change the member or not, probably do want to compare before/after readonly state to avoid `notifyReadOnlyState()` notifying the same state twice in a row if the transition in/out of staging mode doesn't actually change the computed value of `isReadOnly()` (i.e. if the parentContext was already readOnly for some reason, don't want to re-notify readonly)